### PR TITLE
nydus: add raw bootstrap digest annotation for validation

### PR DESCRIPTION
**Pull request by zyfjeff**
Ref: 
https://github.com/containerd/nydus-snapshotter/pull/45

This pr adds an annotation(containerd.io/snapshot/nydus-bootstrap-digest) for raw bootstrap validate

Signed-off-by: zyfjeff <tianqian.zyf@alibaba-inc.com>

### CODE REVIEW

File: pkg/driver/nydus/utils/utils.go [hunk 1]
```diff
@@ -121,3 +146,29 @@ func (c NydusPlatformComparer) Match(platform ocispec.Platform) bool {
...
 	return c.MatchComparer.Less(a, b)
 }
+
+func GetRawBootstrapFromV6(file io.ReadSeeker) (io.Reader, error) {
+	buf := make([]byte, 8)
+	_, err := file.Seek(RafsV6SuppeOffset, io.SeekStart)
+	if err != nil {
+		return nil, fmt.Errorf("invalid bootstrap size, the bootstrap must large than %d", RafsV6SuppeOffset)
+	}
+	file.Read(buf)
```
Comment (imeoer) [inline pkg/driver/nydus/utils/utils.go:156]: The error should be handled?

File: pkg/driver/nydus/utils/utils.go [hunk 2]
```diff
@@ -121,3 +146,29 @@ func (c NydusPlatformComparer) Match(platform ocispec.Platform) bool {
...
+	if err != nil {
+		return nil, fmt.Errorf("invalid bootstrap size, the bootstrap must large than %d", ChunkInfoOffset)
+	}
+
+	file.Read(buf)
+	rawSize := nativeEndian.Uint64(buf)
+	if rawSize < MaxSuperBlockSize {
+		return nil, fmt.Errorf("invalid bootstrap size %d", rawSize)
+	}
+	file.Seek(0, 0)
```
Comment (imeoer) [inline pkg/driver/nydus/utils/utils.go:172]: We'd better use `io.SeekStart` and handle the error.

**Comment by imeoer** [comment]
It's a feature addition for the nydus driver, can we change the commit title like this `nydus: add raw bootstrap digest annotation for validation`?

**Comment by zyfjeff** [comment]
@imeoer ping

File: pkg/driver/nydus/utils/utils.go [hunk 3]
```diff
@@ -121,3 +146,32 @@ func (c NydusPlatformComparer) Match(platform ocispec.Platform) bool {
...
+	if nativeEndian.Uint32(buf) != RafsV6Magic {
+		return nil, fmt.Errorf("invalid bootstrap v6 format")
+	}
+
+	_, err = file.Seek(ChunkInfoOffset, io.SeekStart)
+	if err != nil {
+		return nil, fmt.Errorf("invalid bootstrap size, the bootstrap must large than %d", ChunkInfoOffset)
+	}
+
+	file.Read(buf)
```
Comment (imeoer) [inline pkg/driver/nydus/utils/utils.go:170]: We'd better handle all possible errors.

File: pkg/driver/nydus/utils/utils.go [hunk 4]
```diff
@@ -121,3 +146,32 @@ func (c NydusPlatformComparer) Match(platform ocispec.Platform) bool {
...
+	rawSize := nativeEndian.Uint64(buf)
+	if rawSize < MaxSuperBlockSize {
+		return nil, fmt.Errorf("invalid bootstrap size %d", rawSize)
+	}
+	file.Seek(0, io.SeekStart)
```
Comment (imeoer) [inline pkg/driver/nydus/utils/utils.go:175]: Ditto.

File: pkg/driver/nydus/utils/utils.go [hunk 5]
```diff
@@ -121,3 +146,32 @@ func (c NydusPlatformComparer) Match(platform ocispec.Platform) bool {
...
+		return nil, fmt.Errorf("failed to read magic number")
+	}
+
+	if nativeEndian.Uint32(buf) != RafsV6Magic {
+		return nil, fmt.Errorf("invalid bootstrap v6 format")
+	}
+
+	_, err = file.Seek(ChunkInfoOffset, io.SeekStart)
+	if err != nil {
+		return nil, fmt.Errorf("invalid bootstrap size, the bootstrap must large than %d", ChunkInfoOffset)
```
Comment (imeoer) [inline pkg/driver/nydus/utils/utils.go:167]: Ditto.

File: pkg/driver/nydus/utils/utils.go [hunk 6]
```diff
@@ -121,3 +146,32 @@ func (c NydusPlatformComparer) Match(platform ocispec.Platform) bool {
...
+
+func GetRawBootstrapFromV6(file io.ReadSeeker) (io.Reader, error) {
+	buf := make([]byte, 8)
+	_, err := file.Seek(RafsV6SuppeOffset, io.SeekStart)
+	if err != nil {
+		return nil, fmt.Errorf("invalid bootstrap size, the bootstrap must large than %d", RafsV6SuppeOffset)
+	}
+	_, err = file.Read(buf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read magic number")
```
Comment (imeoer) [inline pkg/driver/nydus/utils/utils.go:158]: Ditto.

File: pkg/driver/nydus/utils/utils.go [hunk 7]
```diff
@@ -121,3 +146,32 @@ func (c NydusPlatformComparer) Match(platform ocispec.Platform) bool {
 func (c NydusPlatformComparer) Less(a, b ocispec.Platform) bool {
 	return c.MatchComparer.Less(a, b)
 }
+
+func GetRawBootstrapFromV6(file io.ReadSeeker) (io.Reader, error) {
+	buf := make([]byte, 8)
+	_, err := file.Seek(RafsV6SuppeOffset, io.SeekStart)
+	if err != nil {
+		return nil, fmt.Errorf("invalid bootstrap size, the bootstrap must large than %d", RafsV6SuppeOffset)
```
Comment (imeoer) [inline pkg/driver/nydus/utils/utils.go:154]: Should use `errors.Wrapf`.

**Comment by zyfjeff** [comment]
@imeoer PTAL, BTW your review comments helped me a lot, thank you very much.

**Review by imeoer — APPROVED** [review]

**Comment by imeoer** [comment]
@zyfjeff Please add SOB for all commits, so that we can merge it.

**Comment by zyfjeff** [comment]
@imeoer SOB already exits

**Comment by zyfjeff** [comment]
@imeoer SOB fixed

**Pull request merged**